contracts/truefi2/libraries/ABDKMath64x64.sol

```diff
@@ -0,0 +1,210 @@
+        if (x & 0x100000 > 0) result = (result * 0x100000000000B17217F7D1D351A389D40) >> 128;
+        if (x & 0x80000 > 0) result = (result * 0x10000000000058B90BFBE8E8B2D3D4EDE) >> 128;
+        if (x & 0x40000 > 0) result = (result * 0x1000000000002C5C85FDF4741BEA6E77E) >> 128;
+        if (x & 0x20000 > 0) result = (result * 0x100000000000162E42FEFA39FE95583C2) >> 128;
+        if (x & 0x10000 > 0) result = (result * 0x1000000000000B17217F7D1CFB72B45E1) >> 128;
+        if (x & 0x8000 > 0) result = (result * 0x100000000000058B90BFBE8E7CC35C3F0) >> 128;
+        if (x & 0x4000 > 0) result = (result * 0x10000000000002C5C85FDF473E242EA38) >> 128;
+        if (x & 0x2000 > 0) result = (result * 0x1000000000000162E42FEFA39F02B772C) >> 128;
+        if (x & 0x1000 > 0) result = (result * 0x10000000000000B17217F7D1CF7D83C1A) >> 128;
+        if (x & 0x800 > 0) result = (result * 0x1000000000000058B90BFBE8E7BDCBE2E) >> 128;
+        if (x & 0x400 > 0) result = (result * 0x100000000000002C5C85FDF473DEA871F) >> 128;
+        if (x & 0x200 > 0) result = (result * 0x10000000000000162E42FEFA39EF44D91) >> 128;
+        if (x & 0x100 > 0) result = (result * 0x100000000000000B17217F7D1CF79E949) >> 128;
+        if (x & 0x80 > 0) result = (result * 0x10000000000000058B90BFBE8E7BCE544) >> 128;
+        if (x & 0x40 > 0) result = (result * 0x1000000000000002C5C85FDF473DE6ECA) >> 128;
+        if (x & 0x20 > 0) result = (result * 0x100000000000000162E42FEFA39EF366F) >> 128;
+        if (x & 0x10 > 0) result = (result * 0x1000000000000000B17217F7D1CF79AFA) >> 128;
+        if (x & 0x8 > 0) result = (result * 0x100000000000000058B90BFBE8E7BCD6D) >> 128;
+        if (x & 0x4 > 0) result = (result * 0x10000000000000002C5C85FDF473DE6B2) >> 128;
+        if (x & 0x2 > 0) result = (result * 0x1000000000000000162E42FEFA39EF358) >> 128;
+        if (x & 0x1 > 0) result = (result * 0x10000000000000000B17217F7D1CF79AB) >> 128;
+
+        result >>= uint256(int256(63 - (x >> 64)));
+        require(result <= uint256(int256(MAX_64x64)));
+
+        return int128(int256(result));
+    }
+
+    /**
+     * Calculate x * y rounding down.  Revert on overflow.
+     *
+     * @param x signed 64.64-bit fixed point number
+     * @param y signed 64.64-bit fixed point number
+     * @return signed 64.64-bit fixed point number
+     */
+    function mul(int128 x, int128 y) internal pure returns (int128) {
+        int256 result = (int256(x) * y) >> 64;
+        require(result >= MIN_64x64 && result <= MAX_64x64);
+        return int128(result);
+    }
+
+    /**
+     * @dev calculate x^y using the fact that
+     * x^y = (2^log2(x))^y = 2^(y * log2(x))
+     *
+     * @param x is 64.64-bit fixed point number
+     * @param y is basis-point precision (10000 = 100%)
+     * @return signed 64.64-bit fixed point number
+     */
+    function pow(int128 x, uint256 y) internal pure returns (int128) {
```

Comment:
1. Let's make both arguments 64.64-bit fixed point numbers. We can handle the `fromUint(y) / 10000` conversion outside of this library, which would make this library agnostic to our contract basis precision.
2. We should also rename the function to something like `powf()`, which disambiguates it from ABDK's original `pow()` function that we removed. I'm not sure `powf()` is the best name since the `f` usually means float, so if you have better suggestions I'm happy to hear them.
3. Let's move this above `log_2()` and below `mul()` for consistency with ABDK.
4. Prepend a comment line to explain TrueFi's modifications on top of ABDK, something like `TFTFTF Add a new pow implementation that takes 64.64-bit fixed point numbers for both arguments.`

Reply:
1. 👍 
2. We can call it `powi()` to keep in line with ABDK's naming
3. 👍 
4. 👍

Reply:
2.
I think the `i` in `muli()`, etc. stands for signed integer, which indicates the second arg:
https://github.com/abdk-consulting/abdk-libraries-solidity/blob/d8817cb600381319992d7caa038bf4faceb1097f/ABDKMath64x64.sol#L163

You can also see that `mulu()` multiplies by an unsigned integer second arg:
https://github.com/abdk-consulting/abdk-libraries-solidity/blob/d8817cb600381319992d7caa038bf4faceb1097f/ABDKMath64x64.sol#L201

Their convention for no suffix is to take an int128 [fixed64x64] as second arg:
https://github.com/abdk-consulting/abdk-libraries-solidity/blob/d8817cb600381319992d7caa038bf4faceb1097f/ABDKMath64x64.sol#L147

Unfortunately ABDK doesn't follow this convention for their `pow()`, which instead takes a uint256 second arg and should probably be named `powi()` instead.

Reply:
Our implementation expects signed integers although it will fail for negative `y`. Let's keep `powi()`. Or name it totally differently, e.g. `decimalPow()`

Reply:
I didn't mean to derail the whole PR for a single function name. Sorry about this :)

The [`powX()` for type `X`] naming convention is an often confusing legacy of ANSI C, so I agree that we might do better just expanding the name. I like `decimalPow()`, although our current implementation (as of dd54f45) is more of a `2^(y * log2(x))`, which doesn't directly deal with decimals (vs binary fixed point).

Let's settle on `fixed64x64Pow()`, which is a little verbose but should be completely unambiguous.